File: windows/src/AutoSwitchKVM.Core/SelectionEngine.cs
using AutoSwitchKVM.Core.Models;

namespace AutoSwitchKVM.Core;

/// Core state machine: maps USB source presence to connect/disconnect of the enabled Bluetooth
/// devices. Ported 1:1 from osx/Sources/AutoSwitchKVM/Engine/SelectionEngine.swift.
///
/// Threading: like the macOS engine (which is @MainActor), this assumes single-threaded use - the
/// app marshals calls onto the UI thread (DispatcherQueue). It does not lock internally.
///
/// Determinism for tests: the production debounce/poll timers are separate from the transition
/// logic. Tests drive HandleUsb/ApplyUsbSnapshot + EvaluateNowAsync directly (no timers), and set
/// SleepHook to a no-op so per-device delay, post-pair settle, and retry backoff are instant.
public sealed class SelectionEngine
{
    private readonly IUsbMonitor _usb;
    private readonly IBluetoothController _bt;
    private readonly Random _rng = new();

    private readonly HashSet<ushort> _presentProductIDs = new();
    /// Presence the automation last acted on (a latch, distinct from Selected which tracks raw
    /// presence even while paused).
    private bool _automationActed;
    private int _runToken;                       // bumps each evaluation to cancel in-flight work
    private readonly HashSet<Guid> _busyDevices = new();
    private readonly Dictionary<Guid, DeviceStatus> _statuses = new();

    private CancellationTokenSource? _debounceCts;

    /// The live configuration. The app may swap this on a profile switch (then call RefreshMonitoring
    /// + ReevaluateAsync). Tests mutate the same instance (e.g. Config.Paused = true).
    public AppConfig Config { get; set; }

    // ---- Observable state (UI binds to these; Changed fires after any update) ----
    public bool Selected { get; private set; }
    public bool BluetoothPowered { get; private set; } = true;
    public string LastReason { get; private set; } = "startup";
    public IReadOnlyDictionary<Guid, DeviceStatus> Statuses => _statuses;
    public event Action? Changed;

    /// Raised on connect/disconnect events for user-facing notices (title, body).
    public Action<string, string>? OnNotice;
    /// Raised when a managed device drops while the source is still present (not an engine action).
    public Action<BTDevice>? OnUnexpectedDisconnect;

    // ---- Test seams (internal; see InternalsVisibleTo) ----
    /// Replaceable delay used for per-device connect delay, post-pair settle, and retry backoff.
    internal Func<int, Task> SleepHook = ms => Task.Delay(ms);
    /// Settle delay after pairing before the clean reconnect (ms).
    internal int PostPairSettleMs = 800;

    public SelectionEngine(AppConfig config, IUsbMonitor usb, IBluetoothController bt)
    {
        Config = config;
        _usb = usb;
        _bt = bt;
    }

    // MARK: - Status monitoring

    /// (Re)register Bluetooth connect/disconnect observers for the current device set.
    public void RefreshMonitoring()
    {
        _bt.StartMonitoring(Config.Devices.Select(d => d.Address), HandleStatusEvent);
    }

    private void HandleStatusEvent(ConnectionChange change)
    {
        var device = Config.Devices.FirstOrDefault(d => d.AddressDigits == change.Address);
        if (device is null) return;
        if (_busyDevices.Contains(device.Id)) return;
        UpdateStatus(device, change.IsConnected);
    }

    /// Poll live connection state for all devices (safety net for a missed event). Refreshes adapter
    /// power first; skips devices that are mid-transition.
    public async Task PollStatusesAsync()
    {
        await RefreshPowerAsync();
        foreach (var device in Config.Devices)
        {
            if (_busyDevices.Contains(device.Id)) continue;
            if (!BluetoothPowered)
            {
                SetStatus(device.Id, DeviceStatus.BluetoothOff);
                continue;
            }
            var connected = await _bt.IsConnectedAsync(device.Address);
            UpdateStatus(device, connected);
        }
    }

    /// Reflect a device's live connection state, and flag a passive drop (connected -> disconnected
    /// while the source is present, BT is on, and we're not mid-transition) as unexpected.
    private void UpdateStatus(BTDevice device, bool connected)
    {
        DeviceStatus? old = _statuses.TryGetValue(device.Id, out var o) ? o : null;
        var @new = connected ? DeviceStatus.Connected : DeviceStatus.Disconnected;
        if (old != @new) SetStatus(device.Id, @new);

        if (!connected && old == DeviceStatus.Connected && Selected && BluetoothPowered && device.Enabled)
        {
            OnUnexpectedDisconnect?.Invoke(device);
        }
    }

    /// Refresh adapter power; on a false->true transition, re-seed so devices reconnect.
    public async Task RefreshPowerAsync()
    {
        var powered = await _bt.IsPoweredOnAsync() ?? true;
        if (powered == BluetoothPowered) return;
        BluetoothPowered = powered;
        RaiseChanged();
        if (powered) await SeedAsync();
    }

    // MARK: - Seeding & USB input

    /// Establish presence from scratch (launch, wake, profile switch) and evaluate.
    public async Task SeedAsync()
    {
        _presentProductIDs.Clear();
        _automationActed = false;
        var source = Config.Source;
        if (source is not null)
        {
            foreach (var d in _usb.Snapshot())
                if (d.VendorID == source.VendorID && source.ProductIDs.Contains(d.ProductID))
                    _presentProductIDs.Add(d.ProductID);
        }
        LastReason = "initial scan";
        await EvaluateNowAsync();
    }

    public Task ReevaluateAsync() => SeedAsync();

    /// Recompute presence from a full USB snapshot (the Windows reconcile path). Pure: call
    /// EvaluateNowAsync (or OnUsbSnapshotAsync) to act on it.
    public void ApplyUsbSnapshot(IReadOnlyList<UsbDeviceInfo> snapshot)
    {
        _presentProductIDs.Clear();
        var source = Config.Source;
        if (source is not null)
        {
            foreach (var d in snapshot)
                if (d.VendorID == source.VendorID && source.ProductIDs.Contains(d.ProductID))
                    _presentProductIDs.Add(d.ProductID);
        }
        LastReason = "usb snapshot";
    }

    /// Incremental USB event (mirrors the macOS handleUSB). Pure: updates presence only.
    public void HandleUsb(ushort vendorId, ushort productId, bool added)
    {
        var source = Config.Source;
        if (source is null || vendorId != source.VendorID || !source.ProductIDs.Contains(productId))
            return;

        if (added)
        {
            _presentProductIDs.Add(productId);
            LastReason = $"source arrived (0x{productId:X4})";
        }
        else
        {
            _presentProductIDs.Remove(productId);
            LastReason = $"source removed (0x{productId:X4})";
        }
    }

    // ---- Production entry points: update presence then debounce-evaluate ----

    public void OnUsbSnapshot(IReadOnlyList<UsbDeviceInfo> snapshot)
    {
        ApplyUsbSnapshot(snapshot);
        ScheduleEvaluate();
    }

    public void OnUsbEvent(ushort vendorId, ushort productId, bool added)
    {
        HandleUsb(vendorId, productId, added);
        ScheduleEvaluate();
    }

    /// Debounced evaluate: arrival debounce when the source is present (snappier connect), departure
    /// debounce when gone. Cancels any pending evaluate.
    public void ScheduleEvaluate()
    {
        _debounceCts?.Cancel();
        var cts = new CancellationTokenSource();
        _debounceCts = cts;
        var ms = _presentProductIDs.Count == 0 ? Config.DebounceMs : Config.ArrivalDebounceMs;
        _ = DebouncedEvaluateAsync(ms, cts.Token);
    }

    private async Task DebouncedEvaluateAsync(int ms, CancellationToken ct)
    {
        try { await Task.Delay(ms, ct); }
        catch (OperationCanceledException) { return; }
        if (ct.IsCancellationRequested) return;
        await EvaluateNowAsync();
    }

    /// The transition logic, awaitable so tests can drive it deterministically without timers.
    public async Task EvaluateNowAsync()
    {
        var present = _presentProductIDs.Count > 0;
        Selected = present;                 // always reflect real presence for the UI
        RaiseChanged();

        if (Config.Paused) return;          // automation suspended; don't act

        if (present && !_automationActed)
        {
            _automationActed = true;
            _runToken++;
            await ConnectAllAsync(_runToken);
        }
        else if (!present && _automationActed)
        {
            _automationActed = false;
            _runToken++;
            await DisconnectAllAsync();
        }
    }

    // MARK: - Manual quick actions (ignore selection, operate on all configured devices)

    public async Task ConnectAllNowAsync()
    {
        _runToken++;
        var token = _runToken;
        foreach (var device in Config.Devices)
        {
            _busyDevices.Add(device.Id);
            await ConnectOneAsync(device, token, respectSelection: false);
            _busyDevices.Remove(device.Id);
        }
    }

    public async Task DisconnectAllNowAsync()
    {
        foreach (var device in Config.Devices)
        {
            _busyDevices.Add(device.Id);
            await DisconnectOneAsync(device);
            _busyDevices.Remove(device.Id);
        }
    }

    // MARK: - Retry backoff

    private const double MaxBackoffSeconds = 30.0;

    /// Backoff (seconds) between connect retries: base * 2^(attempt-1), capped. With base 2 and a
    /// 1-based attempt: 2, 4, 8, 16, 30 (capped). Deterministic.
    public static double BackoffSeconds(int @base, int attempt)
    {
        var value = @base * Math.Pow(2.0, attempt - 1);
        return Math.Min(value, MaxBackoffSeconds);
    }

    /// Backoff in milliseconds with +/-15% jitter (used between connect retries).
    internal int BackoffMillisJittered(int @base, int attempt)
    {
        var jittered = BackoffSeconds(@base, attempt) * (0.85 + _rng.NextDouble() * 0.30);
        return (int)Math.Max(0, jittered * 1000.0);
    }

    // MARK: - Actions

    private async Task ConnectAllAsync(int token)
    {
        foreach (var device in Config.Devices)
        {
            if (!device.Enabled) continue;
            if (token != _runToken) return;
            _busyDevices.Add(device.Id);
            await ConnectOneAsync(device, token, respectSelection: true);
            _busyDevices.Remove(device.Id);
        }
    }

    private async Task ConnectOneAsync(BTDevice device, int token, bool respectSelection)
    {
        var cfg = Config;
        var addr = device.Address;
        SetStatus(device.Id, DeviceStatus.Connecting);

        // Optional per-device stagger before connecting.
        if (device.ConnectDelayMs > 0)
        {
            await SleepHook(device.ConnectDelayMs);
            if (token != _runToken || (respectSelection && !Selected))
            {
                SetStatus(device.Id, DeviceStatus.Idle);
                return;
            }
        }

        var attempt = 0;
        while (attempt < cfg.ConnectRetryMax)
        {
            if (!BluetoothPowered) { SetStatus(device.Id, DeviceStatus.BluetoothOff); return; }
            var aborted = token != _runToken || (respectSelection && !Selected);
            if (aborted) { SetStatus(device.Id, DeviceStatus.Idle); return; }
            attempt++;

            if (!device.ManagePairing && await _bt.IsConnectedAsync(addr))
            {
                SetStatus(device.Id, DeviceStatus.Connected);
                return;
            }

            // A clean connect = ConnectAsync returned without throwing. That's the reliable "device
            // is up" signal; IsConnected alone can be a link-only state that doesn't actually work.
            var connectedCleanly = false;
            try
            {
                if (device.ManagePairing)
                {
                    await WithTimeoutAsync(ct => _bt.PairAsync(addr, ct));
                    // The connection reported right after pairing is unreliable; drop and reconnect.
                    try { await WithTimeoutAsync(ct => _bt.DisconnectAsync(addr, ct)); } catch { }
                    await SleepHook(PostPairSettleMs);
                }
                await WithTimeoutAsync(ct => _bt.ConnectAsync(addr, ct));
                connectedCleanly = true;
            }
            catch (Exception ex)
            {
                SetStatus(device.Id, DeviceStatus.Error(ex.Message));
            }

            if (connectedCleanly && await _bt.IsConnectedAsync(addr))
            {
                SetStatus(device.Id, DeviceStatus.Connected);
                OnNotice?.Invoke("Connected", device.Name);
                return;
            }
            else
            {
                // Don't trust a link-only state; drop any partial connection so the next try is clean.
                try { await WithTimeoutAsync(ct => _bt.DisconnectAsync(addr, ct)); } catch { }
            }

            if (attempt < cfg.ConnectRetryMax)
                await SleepHook(BackoffMillisJittered(cfg.ConnectRetrySecs, attempt));
        }

        if (GetStatus(device.Id) != DeviceStatus.Connected)
        {
            SetStatus(device.Id, DeviceStatus.Error($"gave up after {cfg.ConnectRetryMax} attempts"));
            OnNotice?.Invoke("Connection failed", $"{device.Name}: gave up after {cfg.ConnectRetryMax} attempts");
        }
    }

    private async Task DisconnectAllAsync()
    {
        foreach (var device in Config.Devices)
        {
            if (!device.Enabled) continue;
            _busyDevices.Add(device.Id);
            await DisconnectOneAsync(device);
            _busyDevices.Remove(device.Id);
        }
    }

    private async Task DisconnectOneAsync(BTDevice device)
    {
        var addr = device.Address;
        try
        {
            await WithTimeoutAsync(ct => _bt.DisconnectAsync(addr, ct));
            if (device.ManagePairing)
                await WithTimeoutAsync(ct => _bt.UnpairAsync(addr, ct));
            SetStatus(device.Id, DeviceStatus.Disconnected);
            OnNotice?.Invoke("Disconnected", device.Name);
        }
        catch (Exception ex)
        {
            SetStatus(device.Id, DeviceStatus.Error(ex.Message));
        }
    }

    // MARK: - Helpers

    /// Run a Bluetooth call with the configured per-call timeout.
    private async Task WithTimeoutAsync(Func<CancellationToken, Task> op)
    {
        var secs = Math.Max(1, Config.BtCallTimeoutSecs);
        using var cts = new CancellationTokenSource();
        var opTask = op(cts.Token);
        var delayTask = Task.Delay(TimeSpan.FromSeconds(secs), cts.Token);
        var winner = await Task.WhenAny(opTask, delayTask);
        cts.Cancel();
        if (winner == delayTask && !opTask.IsCompleted)
            throw new TimeoutException($"Bluetooth call timed out after {secs}s");
        await opTask; // observe exceptions / completion
    }

    private DeviceStatus? GetStatus(Guid id) => _statuses.TryGetValue(id, out var s) ? s : null;

    private void SetStatus(Guid id, DeviceStatus status)
    {
        if (_statuses.TryGetValue(id, out var existing) && existing == status) return;
        _statuses[id] = status;
        RaiseChanged();
    }

    private void RaiseChanged() => Changed?.Invoke();
}

File: osx/Sources/AutoSwitchKVM/Engine/SelectionEngine.swift
import Foundation
import Combine

enum DeviceStatus: Equatable {
    case idle
    case connecting
    case connected
    case disconnected
    case bluetoothOff
    case error(String)

    var label: String {
        switch self {
        case .idle: return "Idle"
        case .connecting: return "Connecting…"
        case .connected: return "Connected"
        case .disconnected: return "Disconnected"
        case .bluetoothOff: return "Bluetooth off"
        case .error(let m): return "Error: \(m)"
        }
    }
}

/// Core state machine: maps USB source presence to connect/disconnect of the enabled
/// Bluetooth devices. Mirrors the Hammerspoon prototype (debounce, bounded retries,
/// abort-if-deselected, pair/unpair where configured).
@MainActor
final class SelectionEngine: ObservableObject {
    @Published private(set) var selected: Bool = false
    @Published private(set) var statuses: [UUID: DeviceStatus] = [:]
    @Published private(set) var bluetoothPowered: Bool = true
    @Published var lastReason: String = "startup"

    /// Optional sink for user-facing notices (title, body) raised on connect/disconnect events.
    var onNotice: ((String, String) -> Void)?
    /// Called when a managed device drops while the source is still present (not an engine action).
    var onUnexpectedDisconnect: ((BTDevice) -> Void)?

    private let store: ConfigStore
    private let usb: USBMonitoring
    private let bt: BluetoothController

    private var cancellables = Set<AnyCancellable>()
    private var presentProductIDs = Set<UInt16>()
    /// Presence the automation last connected/disconnected for (latch, distinct from `selected`
    /// which tracks raw presence even while paused).
    private var automationActed = false
    private var debounceTimer: Timer?
    private var runToken = 0   // bumps on every evaluation to cancel in-flight work

    private var statusTimer: Timer?
    /// Devices with an in-flight connect/disconnect; the poller leaves their status alone.
    private var busyDevices = Set<UUID>()

    init(store: ConfigStore, usb: USBMonitoring, bt: BluetoothController) {
        self.store = store
        self.usb = usb
        self.bt = bt

        // Events are delivered on the main queue (USBMonitor dispatches there), so assumeIsolated
        // is safe and lets us call the @MainActor handler from the Combine closure.
        usb.events
            .sink { [weak self] event in MainActor.assumeIsolated { self?.handleUSB(event) } }
            .store(in: &cancellables)

        // Re-register Bluetooth observers whenever the device list changes (debounced so rapid
        // settings edits don't thrash). dropFirst avoids double-registering at startup.
        store.$config
            .map(\.devices)
            .removeDuplicates()
            .debounce(for: .milliseconds(300), scheduler: RunLoop.main)
            .dropFirst()
            .sink { [weak self] _ in MainActor.assumeIsolated { self?.refreshMonitoring() } }
            .store(in: &cancellables)
    }

    /// Starts event-driven status updates (IOBluetooth notifications) plus a slow poll as a
    /// safety net in case an event is ever missed.
    func startStatusMonitor(pollIntervalSeconds: Double = 10.0) {
        refreshMonitoring()
        statusTimer?.invalidate()
        statusTimer = Timer.scheduledTimer(withTimeInterval: pollIntervalSeconds, repeats: true) { [weak self] _ in
            Task { @MainActor in await self?.pollStatuses() }
        }
        Task { @MainActor in await pollStatuses() }
    }

    /// (Re)register IOBluetooth connect/disconnect observers for the current device set.
    func refreshMonitoring() {
        let addresses = store.config.devices.map { $0.normalizedAddress }
        bt.startMonitoring(addresses: addresses) { [weak self] addr, connected in
            self?.handleStatusEvent(addr: addr, connected: connected)
        }
    }

    private func handleStatusEvent(addr: String, connected: Bool) {
        guard let device = store.config.devices.first(where: { $0.normalizedAddress == addr }) else { return }
        guard !busyDevices.contains(device.id) else { return }
        updateStatus(device, connected: connected)
    }

    private func pollStatuses() async {
        await refreshPower()
        for device in store.config.devices {
            guard !busyDevices.contains(device.id) else { continue }
            if !bluetoothPowered {
                if statuses[device.id] != .bluetoothOff { statuses[device.id] = .bluetoothOff }
                continue
            }
            let connected = await bt.isConnected(device.normalizedAddress)
            updateStatus(device, connected: connected)
        }
    }

    /// Reflect a device's live connection state, and flag a passive drop (connected → disconnected
    /// while the source is present, BT is on, and we're not mid-transition) as unexpected.
    private func updateStatus(_ device: BTDevice, connected: Bool) {
        let old = statuses[device.id]
        let new: DeviceStatus = connected ? .connected : .disconnected
        if old != new { statuses[device.id] = new }

        if !connected, old == .connected, selected, bluetoothPowered, device.enabled {
            Log.engine.log("\(device.name) dropped unexpectedly while source present")
            onUnexpectedDisconnect?(device)
        }
    }

    /// Refresh adapter power; on a false→true transition, re-evaluate so devices reconnect.
    /// Internal so tests can drive it.
    func refreshPower() async {
        let powered = await bt.isPoweredOn() ?? true
        guard powered != bluetoothPowered else { return }
        bluetoothPowered = powered
        if powered {
            Log.engine.log("Bluetooth powered on → re-evaluating")
            seed()
        } else {
            Log.engine.log("Bluetooth powered off")
        }
    }

    /// Called at launch and on wake to establish current presence from scratch.
    func seed() {
        let source = store.config.source
        presentProductIDs.removeAll()
        automationActed = false   // re-evaluate from scratch (covers profile switches and wake)
        if let source {
            for dev in usb.attachedDevices where dev.vendorID == source.vendorID && source.productIDs.contains(dev.productID) {
                presentProductIDs.insert(dev.productID)
            }
        }
        lastReason = "initial scan"
        scheduleEvaluate()
    }

    // MARK: - USB handling

    /// Internal (not private) so unit tests can simulate USB events directly.
    func handleUSB(_ event: (vendorID: UInt16, productID: UInt16, added: Bool)) {
        guard let source = store.config.source,
              event.vendorID == source.vendorID,
              source.productIDs.contains(event.productID) else { return }

        if event.added {
            presentProductIDs.insert(event.productID)
            lastReason = String(format: "source arrived (0x%04X)", event.productID)
        } else {
            presentProductIDs.remove(event.productID)
            lastReason = String(format: "source removed (0x%04X)", event.productID)
        }
        scheduleEvaluate()
    }

    private func scheduleEvaluate() {
        debounceTimer?.invalidate()
        // Source present now → arrival debounce (snappier connect); gone → departure debounce.
        let ms = presentProductIDs.isEmpty ? store.config.debounceMs : store.config.arrivalDebounceMs
        let interval = Double(ms) / 1000.0
        debounceTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: false) { [weak self] _ in
            Task { @MainActor in self?.evaluate() }
        }
    }

    private func evaluate() {
        Task { @MainActor in await self.evaluateNow() }
    }

    /// The transition logic, awaitable so tests can drive it deterministically without timers.
    /// Internal for testing; production reaches it via the debounce timer in `evaluate()`.
    func evaluateNow() async {
        let present = !presentProductIDs.isEmpty
        selected = present                       // always reflect real presence for the UI

        guard !store.config.paused else { return }   // automation suspended; don't act

        // `automationActed` latches the presence the automation last acted on, so resuming from
        // pause reconciles to the current state.
        if present && !automationActed {
            automationActed = true
            runToken += 1
            Log.engine.log("source present → connecting enabled devices")
            await connectAll(token: runToken)
        } else if !present && automationActed {
            automationActed = false
            runToken += 1
            Log.engine.log("source absent → disconnecting enabled devices")
            await disconnectAll()
        }
    }

    // MARK: - Manual quick actions (menu) — operate on all configured devices, ignore selection

    func connectAllNow() { Task { @MainActor in await connectAllNowImpl() } }
    func disconnectAllNow() { Task { @MainActor in await disconnectAllNowImpl() } }

    func connectAllNowImpl() async {
        runToken += 1
        let token = runToken
        for device in store.config.devices {
            busyDevices.insert(device.id)
            await connectOne(device, token: token, respectSelection: false)
            busyDevices.remove(device.id)
        }
    }

    func disconnectAllNowImpl() async {
        for device in store.config.devices {
            busyDevices.insert(device.id)
            await disconnectOne(device)
            busyDevices.remove(device.id)
        }
    }

    // MARK: - Retry backoff

    private static let maxBackoffSeconds = 30.0

    /// Backoff (seconds) between connect retries: `base * 2^(attempt-1)`, capped. With the default
    /// base of 2s and a 1-based `attempt`, this is 2, 4, 8, 16, 30… Deterministic.
    static func backoffSeconds(base: Int, attempt: Int) -> Double {
        let value = Double(base) * pow(2.0, Double(attempt) - 1.0)
        return min(value, maxBackoffSeconds)
    }

    /// Backoff with ±15% jitter, in nanoseconds (used between connect retries).
    static func backoffNanos(base: Int, attempt: Int) -> UInt64 {
        let jittered = backoffSeconds(base: base, attempt: attempt) * Double.random(in: 0.85...1.15)
        return UInt64(max(0, jittered) * 1_000_000_000)
    }

    // MARK: - Actions

    private func connectAll(token: Int) async {
        for device in store.config.devices where device.enabled {
            guard token == runToken else { return }
            busyDevices.insert(device.id)
            await connectOne(device, token: token, respectSelection: true)
            busyDevices.remove(device.id)
        }
    }

    private func connectOne(_ device: BTDevice, token: Int, respectSelection: Bool) async {
        let cfg = store.config
        let addr = device.normalizedAddress
        statuses[device.id] = .connecting

        // Optional per-device stagger before connecting.
        if device.connectDelayMs > 0 {
            try? await Task.sleep(nanoseconds: UInt64(device.connectDelayMs) * 1_000_000)
            guard token == runToken, (!respectSelection || selected) else {
                statuses[device.id] = .idle; return
            }
        }

        var attempt = 0
        while attempt < cfg.connectRetryMax {
            if !bluetoothPowered { statuses[device.id] = .bluetoothOff; return }
            let aborted = token != runToken || (respectSelection && !selected)
            guard !aborted else { statuses[device.id] = .idle; return }
            attempt += 1
            Log.engine.log("connect \(device.name): attempt \(attempt)/\(cfg.connectRetryMax) (managePairing=\(device.managePairing))")

            if !device.managePairing, await bt.isConnected(addr) {
                Log.bluetooth.log("\(device.name) already connected")
                statuses[device.id] = .connected
                return
            }

            // A clean connect = `connect` returned without throwing, i.e. IOBluetooth's
            // `connectionComplete` fired with success. That's the reliable "HID is up" signal;
            // `isConnected()` alone can be a link-only (ACL) state that doesn't actually work.
            var connectedCleanly = false
            do {
                if device.managePairing {
                    Log.bluetooth.log("pair \(device.name)…")
                    try await withTimeout(cfg.btCallTimeoutSecs) { try await self.bt.pair(addr) }
                    Log.bluetooth.log("pair \(device.name): ok")
                    // The connection IOBluetooth reports right after pairing is unreliable; drop it
                    // and reconnect cleanly.
                    try? await withTimeout(cfg.btCallTimeoutSecs) { try await self.bt.disconnect(addr) }
                    try? await Task.sleep(nanoseconds: 800_000_000)
                    Log.bluetooth.log("post-pair settle done; reconnecting \(device.name)")
                }
                Log.bluetooth.log("openConnection \(device.name)…")
                try await withTimeout(cfg.btCallTimeoutSecs) { try await self.bt.connect(addr) }
                connectedCleanly = true
                Log.bluetooth.log("openConnection \(device.name): completed cleanly")
            } catch {
                statuses[device.id] = .error(String(describing: error))
                Log.bluetooth.error("connect \(device.name): \(error)")
            }

            if connectedCleanly, await bt.isConnected(addr) {
                Log.bluetooth.log("\(device.name) connected ✓")
                statuses[device.id] = .connected
                onNotice?("Connected", device.name)
                return
            } else {
                // Don't trust a link-only state; drop any partial connection so the next attempt is clean.
                Log.bluetooth.log("\(device.name) not cleanly connected after attempt \(attempt); dropping partial link")
                try? await withTimeout(cfg.btCallTimeoutSecs) { try await self.bt.disconnect(addr) }
            }

            if attempt < cfg.connectRetryMax {
                try? await Task.sleep(nanoseconds: Self.backoffNanos(base: cfg.connectRetrySecs, attempt: attempt))
            }
        }
        if case .connected = statuses[device.id] ?? .idle {} else {
            statuses[device.id] = .error("gave up after \(cfg.connectRetryMax) attempts")
            onNotice?("Connection failed", "\(device.name): gave up after \(cfg.connectRetryMax) attempts")
        }
    }

    private func disconnectAll() async {
        for device in store.config.devices where device.enabled {
            busyDevices.insert(device.id)
            await disconnectOne(device)
            busyDevices.remove(device.id)
        }
    }

    private func disconnectOne(_ device: BTDevice) async {
        let cfg = store.config
        let addr = device.normalizedAddress
        do {
            try await withTimeout(cfg.btCallTimeoutSecs) { try await self.bt.disconnect(addr) }
            if device.managePairing {
                try await withTimeout(cfg.btCallTimeoutSecs) { try await self.bt.unpair(addr) }
            }
            statuses[device.id] = .disconnected
            onNotice?("Disconnected", device.name)
        } catch {
            statuses[device.id] = .error(String(describing: error))
            Log.bluetooth.error("disconnect \(device.name): \(error)")
        }
    }

    /// Used by the menu/Settings to force a re-evaluation (e.g. after editing devices).
    func reevaluate() {
        seed()
    }
}
